Respect GRUB2_SEARCH_ROOT_COMMAND

jsmeix opened the pull request:
* Type: **Enhancement**

* Impact: **Normal**

* Reference to related issue (URL):
https://github.com/rear/rear/issues/3024#issuecomment-1630963778

* How was this pull request tested?
Not yet tested

* Brief description of the changes in this pull request:

In the create_grub2_cfg function in lib/bootloader-functions.sh
provide final power to the user and always respect
GRUB2_SEARCH_ROOT_COMMAND when the user has set it, cf.
https://github.com/rear/rear/issues/3024#issuecomment-1630963778

jsmeix commented:
Found same kind of issue against "final power to the user"
in output/USB/Linux-i386/300_create_grub.sh
by hardcoded setting of GRUB2_SEARCH_ROOT_COMMAND
regardless what the user may have specified.

pcahyna commented:
The code being changed misses the intent of the code before #2662 . `set root=cd0` and `search --no-floppy --set=root --file /boot/efiboot.img` should be both executed (`set root=cd0` sets a sensible default that works in most cases and applies when `search --no-floppy --set=root --file /boot/efiboot.img` fails to do its job properly), but right now one executes only one of them ( with `set root=cd0` taking precedence), which fails as soon as the DVD is not in the first DVD drive.

schlomo commented on usr/share/rear/lib/bootloader-functions.sh:
```diff
@@ -547,11 +547,16 @@
+    # cf. https://github.com/rear/rear/issues/3024#issuecomment-1630963778
     if ! test "$grub2_search_root_command" ; then
-        test "$grub2_set_root" && grub2_search_root_command="set root=$grub2_set_root"
+        test "$GRUB2_SEARCH_ROOT_COMMAND" && grub2_search_root_command="$GRUB2_SEARCH_ROOT_COMMAND"
```
Just a thought:

I agree with the check that having the same variable in all caps and in lowercase is at least confusing (although the check is in general wrong as it doesn't know about `default.conf`).

Is that something we can rectify or would that be too much of a change? This file is hard to read in any case, so maybe this is the least of all problems.

@jsmeix you spend more time here than me so I'll defer to your judgement.

jsmeix replied:
Those lowercase named variables are local variables that
intentionally match uppercase named user config variables.
Those lowercase named variables are meant to be used
inside a function or inside a script
when the value could be modified for that local usage
without modifying the user config variable.

I do not like it when user config variables get modified.

I think normally user config variables should be kept as is
so that different scripts could behave according to what
the user had specified and not according to what a previous
script may have changed from what the user had specified.

Probably I am thinking too complicated and I should
just use user config variables like any other variables.

The whole create_grub2_cfg() usage is such a mess
which at least I cannot maintain with reasonable effort
so that I may even clean that up mecilessly right now
and afterwards I can fix what I may have done wrong.

jsmeix replied:
I won't do a general cleanup via this pull request,
cf. https://github.com/rear/rear/pull/3025#issuecomment-1635340530

jsmeix replied:
The lower case variables grub2_set_root_command
and grub2_search_root_command are no longer needed, see
https://github.com/rear/rear/pull/3025#issuecomment-1638156038

schlomo commented on usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh:
```diff
@@ -95,8 +95,12 @@ EOF
+            # we can provide final power to the user and always respect
+            # GRUB2_SEARCH_ROOT_COMMAND when the user has set it
+            # cf. https://github.com/rear/rear/issues/3024#issuecomment-1630963778
+            test "$GRUB2_SEARCH_ROOT_COMMAND" || GRUB2_SEARCH_ROOT_COMMAND="search --no-floppy --set=root --label $efi_label"
```
Shouldn't that be `contains_visible_char` (same for all the others)?

jsmeix replied:
I won't do a general cleanup via this pull request,
cf. https://github.com/rear/rear/pull/3025#issuecomment-1635340530

schlomo commented on usr/share/rear/output/USB/Linux-i386/300_create_grub.sh:
```diff
@@ -79,6 +79,10 @@
+# GRUB2_SEARCH_ROOT_COMMAND when the user has set it
+# cf. https://github.com/rear/rear/issues/3024#issuecomment-1630963778
+test "$GRUB2_SEARCH_ROOT_COMMAND" || GRUB2_SEARCH_ROOT_COMMAND="search --no-floppy --set=root --label $USB_DEVICE_BOOT_LABEL"
+create_grub2_cfg /$USB_PREFIX/kernel /$USB_PREFIX/$REAR_INITRD_FILENAME > $usb_boot_dir/$grub_cfg
```
Indeed, can we be sure that `$USB_PREFIX` or `$REAR_INITRD_FILENAME` never require quoting?

jsmeix replied:
I won't do a general cleanup via this pull request,
cf. https://github.com/rear/rear/pull/3025#issuecomment-1635340530

jsmeix replied:
With current code USB_PREFIX and REAR_INITRD_FILENAME do not require quoting:

```
# find usr/sbin/rear usr/share/rear -type f | xargs grep 'USB_PREFIX=' | grep -v ': *#'

usr/share/rear/prep/USB/default/060_set_usb_device.sh:    USB_PREFIX="rear/$HOSTNAME/$USB_SUFFIX"
usr/share/rear/prep/USB/default/060_set_usb_device.sh:    USB_PREFIX="rear/$HOSTNAME/$( date +%Y%m%d.%H%M )"
usr/share/rear/prep/USB/default/060_set_usb_device.sh:test "$USB_PREFIX" || USB_PREFIX="rear/$HOSTNAME/$(date +%Y%m%d.%H%M)"
```

```
# find usr/sbin/rear usr/share/rear -type f | xargs grep 'REAR_INITRD_FILENAME=' | grep -v ': *#'

usr/share/rear/pack/GNU/Linux/900_create_initramfs.sh:            REAR_INITRD_FILENAME=$VM_UID".initrd"
usr/share/rear/pack/GNU/Linux/900_create_initramfs.sh:            REAR_INITRD_FILENAME="initrd.lz4"
usr/share/rear/pack/GNU/Linux/900_create_initramfs.sh:            REAR_INITRD_FILENAME=$VM_UID".initrd"
usr/share/rear/pack/GNU/Linux/900_create_initramfs.sh:            REAR_INITRD_FILENAME="initrd.xz"
usr/share/rear/pack/GNU/Linux/900_create_initramfs.sh:            REAR_INITRD_FILENAME=$VM_UID".initrd"
usr/share/rear/pack/GNU/Linux/900_create_initramfs.sh:            REAR_INITRD_FILENAME="initrd.cgz"
usr/share/rear/pack/GNU/Linux/900_create_initramfs.sh:            REAR_INITRD_FILENAME=$VM_UID".initrd"
usr/share/rear/pack/GNU/Linux/900_create_initramfs.sh:            REAR_INITRD_FILENAME="initrd.cgz"
usr/share/rear/pack/GNU/Linux/900_create_initramfs.sh:            REAR_INITRD_FILENAME=$VM_UID".initrd"
usr/share/rear/pack/GNU/Linux/900_create_initramfs.sh:            REAR_INITRD_FILENAME="initrd.cgz"
```
VM_UID is only used when $ARCH is "Linux-s390"
and `$VM_UID` is used unquoted in current code.

I won't fix possible general issues in our "Linux-s390" code
via this pull request.

schlomo reviewed (approved):
Unfortunately I cannot comment on the functional side of things and assume it is somehow tested.

Thanks a lot for improving ReaR in this difficult corner of our code!

jsmeix commented:
With the current state of this pull request up to
https://github.com/rear/rear/pull/3025/commits/28fdbd8c7e1d5604ede1c7a73644e0e428c456ce
I works well for my pesonal test case
(which does not mean much because I am not an expert in this area):

My original system is a KVM/QEMU VM
with OVMF "TianoCore" UEFI firmware:
```
# lsblk -ipo NAME,KNAME,PKNAME,TRAN,TYPE,FSTYPE,LABEL,SIZE,MOUNTPOINT
NAME        KNAME     PKNAME   TRAN   TYPE FSTYPE  LABEL                             SIZE MOUNTPOINT
/dev/sr0    /dev/sr0           sata   rom  iso9660 SLE-15-SP4-Full-x86_64151.11.001 12.1G 
/dev/vda    /dev/vda                  disk                                            15G 
|-/dev/vda1 /dev/vda1 /dev/vda        part vfat                                      512M /boot/efi
|-/dev/vda2 /dev/vda2 /dev/vda        part ext4                                     12.5G /
`-/dev/vda3 /dev/vda3 /dev/vda        part swap                                        2G [SWAP]

# grep -v '^$' etc/rear/local.conf 
FIRMWARE_FILES=( 'no' )
MODULES=( 'loaded_modules' )
PROGRESS_MODE="plain"
PROGRESS_WAIT_SECONDS="5"
SSH_ROOT_PASSWORD="rear"
USE_DHCLIENT="yes"
USE_SERIAL_CONSOLE="no"
OUTPUT=ISO
BACKUP=NETFS
BACKUP_OPTIONS="nfsvers=3,nolock"
BACKUP_URL=nfs://192.168.122.1/nfs

# usr/sbin/rear -D mkrescue
...
Found EFI system partition /dev/vda1 on /boot/efi type vfat
Using UEFI Boot Loader for Linux (USING_UEFI_BOOTLOADER=1)
Using '/usr/bin/xorrisofs' to create ISO filesystem images
Using autodetected kernel '/boot/vmlinuz-5.14.21-150400.22-default' as kernel in the recovery system
...
Trying to find what to use as UEFI bootloader...
Trying to find a 'well known file' to be used as UEFI bootloader...
Using '/boot/efi/EFI/sles/grubx64.efi' as UEFI bootloader file
...
Let GRUB2 load kernel /isolinux/kernel
Let GRUB2 load initrd /isolinux/initrd.cgz
Set GRUB2 default root device via 'set root=cd0'
Let GRUB2 search root device via 'search --no-floppy --set=root --file /boot/efiboot.img'
No serial console in GRUB2 (USE_SERIAL_CONSOLE is not true)
GRUB2 modules to load: ext2 fat part_gpt part_msdos
Making ISO image
Wrote ISO image: /root/rear/var/lib/rear/output/rear-localhost.iso (186M)
...

# mount -o loop /root/rear/var/lib/rear/output/rear-localhost.iso /tmp/iso

# cat /tmp/iso/boot/grub/grub.cfg
set root=cd0
search --no-floppy --set=root --file /boot/efiboot.img
echo "Using root device $root from those available devices:"
ls
insmod all_video
set gfxpayload=keep
insmod part_gpt
insmod part_msdos
insmod ext2
echo 'No serial console (USE_SERIAL_CONSOLE was not true)'
set timeout="300"
set default="chainloader"
set fallback="chainloader"
echo 'Switching to GRUB boot menu...'
sleep --verbose --interruptible 30
menuentry "Relax-and-Recover (BIOS or UEFI without Secure Boot)" --id=rear {
    insmod gzio
    insmod xzio
    echo 'Loading kernel /isolinux/kernel ...'
    linux /isolinux/kernel root=UUID=7620c942-d8c0-45a8-9f8f-241e02f6f0c0  selinux=0
    echo 'Loading initial ramdisk /isolinux/initrd.cgz ...'
    initrd /isolinux/initrd.cgz
}
menuentry "Relax-and-Recover (UEFI and Secure Boot)" --id=rear_secure_boot {
    insmod gzio
    insmod xzio
    echo 'Loading kernel /isolinux/kernel ...'
    linuxefi /isolinux/kernel root=UUID=7620c942-d8c0-45a8-9f8f-241e02f6f0c0  selinux=0
    echo 'Loading initial ramdisk /isolinux/initrd.cgz ...'
    initrdefi /isolinux/initrd.cgz
}
menuentry "Boot next EFI" --id=chainloader {
    insmod chain
    search --fs-uuid --no-floppy --set=esp 8AD9-1260
    chainloader ($esp)/EFI/sles/grubx64.efi
}
menuentry "Reboot" --id=reboot {
    reboot
}
menuentry "Exit to EFI shell" --id=exit {
    exit
}
```
excerpts from var/log/rear/rear-localhost.log
```
+ source /root/rear/usr/share/rear/output/ISO/Linux-i386/250_populate_efibootimg.sh
...
++ cp -v /boot/efi/EFI/sles/grubx64.efi /var/tmp/rear.SEIjjZqoarZFztm/tmp/mnt/EFI/BOOT/BOOTX64.efi
...
++ GRUB2_SET_ROOT_COMMAND='set root=cd0'
...
++ GRUB2_SEARCH_ROOT_COMMAND='search --no-floppy --set=root --file /boot/efiboot.img'
++ create_grub2_cfg /isolinux/kernel /isolinux/initrd.cgz
...
++ grub2-mkstandalone -v '--modules=ext2 fat part_gpt part_msdos' -O x86_64-efi -o /var/tmp/rear.SEIjjZqoarZFztm/tmp/mnt/EFI/BOOT/BOOTX64.efi /boot/grub/grub.cfg=/var/tmp/rear.SEIjjZqoarZFztm/tmp/mnt/EFI/BOOT/grub.cfg
```

Using the resulting ISO file in a virtual CDROM
on another same KVM/QEMU VM with OVMF "TianoCore" UEFI firmware
boots the ReaR recovery system for me.

I also did a test with "rear mkbackup" and "rear recover"
which worked for me.

jsmeix commented:
![Screenshot from 2023-07-13 13-11-20](https://github.com/rear/rear/assets/1788608/f576d7cb-31bf-4468-9269-3bfeb5903d9d)

What I personally like most is that now
GRUB2 shows which 'root' device it uses
and which devices are available.

Hopefully this avoids lengthy confusions as in
https://github.com/rear/rear/issues/3024
until finally (thanks to @pcahyna who found it out)
it became clear what the actual issue was, see
https://github.com/rear/rear/issues/3024#issuecomment-1630706633

jsmeix commented:
Tomorrow I will test how it behaves
with two CDROM drives on the replacement system.

jsmeix commented:
First things first:
I will not clean up the whole create_grub2_cfg() usage
via this pull request.
Via this pull request the create_grub2_cfg() usage
should only get fixed where it was actually wrong.
The implementation of the fix introduces the additional
GRUB2_SET_ROOT_COMMAND config variable to get consistency
regarding how GRUB2 sets and/or searches its 'root' device.
A general cleanup of the whole create_grub2_cfg() usage
should be done via a subsequent separated pull request.

jsmeix commented:
With my recent changes here up to
https://github.com/rear/rear/pull/3025/commits/b1d80997dcb083d632f2c8ec6db0e4cc306b41d9
it even works reasonably well with Secure Boot for me:

Same original system as above
but with Secure Boot enabled:
```
# grep -v '^$' etc/rear/local.conf 
FIRMWARE_FILES=( 'no' )
MODULES=( 'loaded_modules' )
PROGRESS_MODE="plain"
PROGRESS_WAIT_SECONDS="5"
SSH_ROOT_PASSWORD="rear"
USE_DHCLIENT="yes"
USE_SERIAL_CONSOLE="no"
SECURE_BOOT_BOOTLOADER="/boot/efi/EFI/sles/shim.efi"
OUTPUT=ISO
BACKUP=NETFS
BACKUP_OPTIONS="nfsvers=3,nolock"
BACKUP_URL=nfs://192.168.122.1/nfs

# usr/sbin/rear -D mkrescue
...
Found EFI system partition /dev/vda1 on /boot/efi type vfat
Using UEFI Boot Loader for Linux (USING_UEFI_BOOTLOADER=1)
Using '/usr/bin/xorrisofs' to create ISO filesystem images
...
Using '/boot/efi/EFI/sles/shim.efi' as UEFI Secure Boot bootloader file
...
Using Shim '/boot/efi/EFI/sles/shim.efi' as first stage UEFI bootloader BOOTX64.efi
Using second stage UEFI bootloader files for Shim: /boot/efi/EFI/sles/grub.efi /boot/efi/EFI/sles/grubx64.efi
Let GRUB2 load kernel /isolinux/kernel
Let GRUB2 load initrd /isolinux/initrd.cgz
Set GRUB2 default root device via 'set root=cd0'
Let GRUB2 search root device via 'search --no-floppy --set=root --file /boot/efiboot.img'
No serial console in GRUB2 (USE_SERIAL_CONSOLE is not true)
Making ISO image
Wrote ISO image: /root/rear/var/lib/rear/output/rear-localhost.iso (180M)
...

# mount -o loop /root/rear/var/lib/rear/output/rear-localhost.iso /tmp/iso
mount: /tmp/iso: WARNING: source write-protected, mounted read-only.

# mount /tmp/iso/boot/efiboot.img /tmp/fat

# find /tmp/iso /tmp/fat -ls | grep -v locale | grep '\.efi'
     1606    913 -rwx------   1 root     root        934680 Jul 14 14:52 /tmp/iso/EFI/BOOT/BOOTX64.efi
     1617   1206 -rwx------   1 root     root       1234944 Jul 14 14:52 /tmp/iso/EFI/BOOT/grub.efi
     1620    140 -rwx------   1 root     root        143360 Jul 14 14:52 /tmp/iso/EFI/BOOT/grubx64.efi
       85    914 -rwxr-xr-x   1 root     root        934680 Jul 14 14:52 /tmp/fat/EFI/BOOT/BOOTX64.efi
       86   1206 -rwxr-xr-x   1 root     root       1234944 Jul 14 14:52 /tmp/fat/EFI/BOOT/grub.efi
       87    140 -rwxr-xr-x   1 root     root        143360 Jul 14 14:52 /tmp/fat/EFI/BOOT/grubx64.efi

# ls -l /boot/efi/EFI/sles/shim.efi /boot/efi/EFI/sles/grub.efi /boot/efi/EFI/sles/grubx64.efi
-rwxr-xr-x 1 root root 1234944 Jul 12 15:12 /boot/efi/EFI/sles/grub.efi
-rwxr-xr-x 1 root root  143360 Jul 12 15:12 /boot/efi/EFI/sles/grubx64.efi
-rwxr-xr-x 1 root root  934680 Jul 12 15:12 /boot/efi/EFI/sles/shim.efi
```

I also did "rear mkbackup".

Same replacement system as above
but with Secure Boot enabled:
```
RESCUE localhost:~ # rear -D recover
...
Running 'finalize' stage ======================
...
Running mkinitrd...
Recreated initrd (/sbin/mkinitrd).
Creating EFI Boot Manager entries...
Creating  EFI Boot Manager entry 'SUSE_LINUX 15.4' for 'EFI\sles\shim.efi' (UEFI_BOOTLOADER='/boot/efi/EFI/sles/shim.efi') 
Installing secure boot loader (shim)...
...
```

When rebooting the recreated system
with the ReaR ISO still in its CDROM drive it booted from the ISO
but then via the default GRUB menu entry `Boot next EFI`
it loaded my SLES15 SP4 system but then that aborted to boot
with some "system is compromised" Secure Boot message
so I removed the ISO from the CDROM drive so that
it boots the recreated system directly from its harddisk
and that "just worked" for me (in Secure Boot mode).

jsmeix commented:
I did some generic cleanup of the create_grub2_cfg function usage
with my recent commits up to
https://github.com/rear/rear/pull/3025/commits/c5eb197aa6f3aa88cd2a98874e56bdb2277ee21b
in particular with
https://github.com/rear/rear/pull/3025/commits/e368c4200cce9bfb4bf60d0862497ff467282304
create_grub2_cfg() no longer accepts a third argument
that could be a "search GRUB2 'root' device command" string
but uses GRUB2_SET_ROOT_COMMAND and GRUB2_SEARCH_ROOT_COMMAND
directly as usual with other config variables.
Therefore the lower case variables
`grub2_set_root_command` and `grub2_search_root_command`
are no longer needed, cf.
https://github.com/rear/rear/pull/3025#discussion_r1262376052

jsmeix commented:
@pcahyna @rear/contributors 
as far as I can imagine currently I think I amd now done
with fixing the create_grub2_cfg function usage
to the extent what I could reasonably do
within one single pull request.

If you like you may have a look at my current changes.
Perhaps you may spot some obvious mistake.

Tomorrow I will redo my above tests with my latest changes.

Ideally - if all goes well - I would like to merge it
later this week.

jsmeix commented:
I did my above tests with my latest changes
and my tests worked again as before.

Two remarks:

Regarding
https://github.com/rear/rear/pull/3025#issuecomment-1635876186
```
When rebooting the recreated system
with the ReaR ISO still in its CDROM drive it booted from the ISO
but then via the default GRUB menu entry Boot next EFI
it loaded my SLES15 SP4 system but then that aborted to boot
with some "system is compromised" Secure Boot message
so I removed the ISO from the CDROM drive so that
it boots the recreated system directly from its harddisk
and that "just worked" for me (in Secure Boot mode).
```
It also works to set in the OVMF "TianoCore" UEFI firmware
to use the harddisk as first boot device.

This time I made a screenshot how it looks when
it shows the "system is compromised" Secure Boot message:

![Screenshot from 2023-07-18 10-19-06](https://github.com/rear/rear/assets/1788608/659920dd-324c-4023-94f6-5b6c4cbaf47c)

jsmeix commented:
The second remark is about the
reinstalled bootloader during "rear recover":

What one gets after reboot after "rear recover"
are those two boot options
```
SLES 15-SP4
Advanced options for SLES 15-SP4
```
(see the above screenshot).

What is not there after "rear recover" is
```
UEFI Firmware Settings
```
which is normally there on SLES 15-SP4 which leads
into the OVMF "TianoCore" UEFI firmware setup menu.

To get the `UEFI Firmware Settings` entry
I did run the YaST bootloader module,
therein I changed an unimportant setting
(I changed in "bootloader options" the timeout)
so that YaST did actually reinstall the bootloader.
Now I get again the usual SLES 15-SP4 bootloader screen:

![Screenshot from 2023-07-18 10-24-16](https://github.com/rear/rear/assets/1788608/8f79b6fe-fdb8-4650-8641-b960bd2146bc)

jsmeix commented:
@pcahyna 
all looks OK now from my point of view with my tests.
I would much appreciate it if you could review my changes.

jsmeix commented:
I also tested
```
OUTPUT=USB
BACKUP=NETFS
BACKUP_URL=usb:///dev/disk/by-label/REAR-000
```
on my above original system (UEFI with Secure Boot).

As far as I see Secure Boot of the ReaR recovery system
is not yet supported for OUTPUT=USB
but UEFI without Secure Boot works for me.

So I had to switch off Secure Boot on my replacement system
to get the ReaR recovery system booted.
I did "rear recover" and rebooted to verify
that the recreated system works.

Then I shut down the replacement system and switched on again
Secure Boot in its OVMF TianoCore UEFI firmware setup menu
and booted it which also works, i.e. the recreated system
also boots with Secure Boot.
So "rear recover" did "the right thing" to also reinstall Shim,
cf. usr/share/rear/finalize/SUSE_LINUX/i386/675_install_shim.sh
in my case because I use SLES 15-SP4.
I didn't test if "rear recover" reinstalls Shim also
for other Linux distributions.

pcahyna commented:
> @pcahyna all looks OK now from my point of view with my tests. I would much appreciate it if you could review my changes.

@jsmeix looking ...

pcahyna commented:
Looks good. The only potentially questionable part is the use of
```
contains_visible_char "$GRUB2_SET_ROOT_COMMAND"
contains_visible_char "$GRUB2_SEARCH_ROOT_COMMAND"
```
which, IIUC, prevents setting the variables to e.g. `" "` as a way to avoid using any set/search command at all.
As an example for another option, I have recently recommended the use of `MKFS_XFS_OPTIONS=" "`  as a way to disable all XFS options (and thus using mkfs.xfs defaults), which is very useful in some situations.

jsmeix commented:
@pcahyna 
thank you for your review!

Regarding questionable 'contains_visible_char' usage:

I introduced it because of
https://github.com/rear/rear/pull/3025#discussion_r1260972359

First I thought it would be against "final power to the user"
if the user could not specify to not use any set/search command.

But then I tested in etc/rear/local.conf
```
GRUB2_SET_ROOT_COMMAND="echo 'No set root command'"
GRUB2_SEARCH_ROOT_COMMAND="echo 'No search root command'"
```
which works for me to get the GRUB2 default 'memdisk'
but this cannot boot the recovery system
(because its kernel and initrd is not on 'memdisk').

Because it cannot work without any set/search command
I think the 'contains_visible_char' usage is OK here
to avoid that things don't work.

What is still questionable is if ReaR should
try to silently "correct" a user specified value
like "only blank characters" == "no characters",
cf. "sequence of zero bytes" == "end of file" in
https://github.com/rear/rear/pull/3027#issuecomment-1641940634
or
if ReaR should not try to "correct" user specified values?

jsmeix commented:
@rear/contributors 
I will merge it in about one hour today
unless objections appear.

pcahyna reviewed (approved):
Yes, setting the variables to some harmless echo commands is a good workaround.
